File: javascript/variable-length-quantity/variable-length-quantity.js
//
// This is only a SKELETON file for the 'Variable Length Quantity' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const encode = (numbers) => {
  const bytes = [];
  for (const number of numbers) {
    if (number === 0) {
      bytes.push(0);
      continue;
    }
    const groups = [];
    let n = number >>> 0; // ensure unsigned
    // Use division and modulo to avoid bitwise issues
    while (n > 0) {
      groups.unshift(n % 128);
      n = Math.floor(n / 128);
    }
    for (let i = 0; i < groups.length - 1; i++) {
      bytes.push(groups[i] | 0x80); // set MSB for all but last
    }
    bytes.push(groups[groups.length - 1]); // last byte, MSB=0
  }
  return bytes;
};

export const decode = (bytes) => {
  const numbers = [];
  let current = 0;
  for (let i = 0; i < bytes.length; i++) {
    const byte = bytes[i];
    current = current * 128 + (byte & 0x7F);
    if ((byte & 0x80) === 0) {
      // convert to unsigned 32-bit explicitly
      current = current >>> 0;
      numbers.push(current);
      current = 0;
    }
  }
  if ((bytes.length > 0) && ((bytes[bytes.length - 1] & 0x80) !== 0)) {
    throw new Error('Incomplete sequence');
  }
  return numbers;
};

File: python/variable-length-quantity/variable_length_quantity.py
def encode(numbers):
    encoded_bytes = []
    for number in numbers:
        if number == 0:
            encoded_bytes.append(0)
            continue
        groups = []
        while number > 0:
            groups.append(number & 0x7F)
            number >>= 7
        for i in range(len(groups) - 1, -1, -1):
            byte = groups[i]
            if i != 0:
                byte |= 0x80  # set continuation bit
            encoded_bytes.append(byte)
    return encoded_bytes


def decode(bytes_):
    numbers = []
    value = 0
    in_sequence = False
    for byte in bytes_:
        value = (value << 7) | (byte & 0x7F)
        if byte & 0x80:
            in_sequence = True
        else:
            numbers.append(value)
            value = 0
            in_sequence = False
    if in_sequence:
        raise ValueError("incomplete sequence")
    return numbers
